crypto/dilithium/p_pqdsa_asn1.c

```diff
@@ -0,0 +1,221 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/evp.h>
+
+#include <openssl/bytestring.h>
+#include <openssl/err.h>
+#include <openssl/mem.h>
+
+#include "../evp_extra/internal.h"
+#include "../fipsmodule/evp/internal.h"
+#include "../internal.h"
+#include "ml_dsa.h"
+#include "internal.h"
+
+static void pqdsa_free(EVP_PKEY *pkey) {
+  PQDSA_KEY_free(pkey->pkey.pqdsa_key);
+  pkey->pkey.pqdsa_key = NULL;
+}
+
+static int pqdsa_get_priv_raw(const EVP_PKEY *pkey, uint8_t *out,
+                                   size_t *out_len) {
+  if (pkey->pkey.pqdsa_key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+    return 0;
+  }
+
+  PQDSA_KEY *key = pkey->pkey.pqdsa_key;
+  const PQDSA *pqdsa = key->pqdsa;
+
+  if (key->private_key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+    return 0;
+  }
+
+  if (pqdsa == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+    return 0;
+  }
+
+  if (out == NULL) {
+    *out_len = key->pqdsa->private_key_len;
+    return 1;
+  }
+
+  if (*out_len < key->pqdsa->private_key_len) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  OPENSSL_memcpy(out, key->private_key, pqdsa->private_key_len);
+  *out_len = pqdsa->private_key_len;
+  return 1;
+}
+
+static int pqdsa_get_pub_raw(const EVP_PKEY *pkey, uint8_t *out,
+                                  size_t *out_len) {
+  if (pkey->pkey.pqdsa_key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+    return 0;
+  }
+
+  PQDSA_KEY *key = pkey->pkey.pqdsa_key;
+  const PQDSA *pqdsa = key->pqdsa;
+
+  if (pqdsa == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+    return 0;
+  }
+
+  if (key->public_key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
+    return 0;
+  }
+
+  if (out == NULL) {
+    *out_len = pqdsa->public_key_len;
+    return 1;
+  }
+
+  if (*out_len < key->pqdsa->public_key_len) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  OPENSSL_memcpy(out, key->public_key, pqdsa->public_key_len);
+  *out_len = pqdsa->public_key_len;
+  return 1;
+}
+
+static int pqdsa_pub_decode(EVP_PKEY *out, CBS *params, CBS *key) {
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 4.
+  // The parameters must be omitted.
+
+  // the only parameter that can be included is the OID which has length 9
+  if (CBS_len(params) != 9) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+  // set the pqdsa params on the fresh pkey
+  EVP_PKEY_pqdsa_set_params(out, OBJ_cbs2nid(params));
```

Comment:
check the return value?

Reply:
woops! checked in [fb8631e](https://github.com/aws/aws-lc/pull/1963/commits/fb8631e132f89e65bda5372f7e76c157202d7341)